stan/math/rev/functor/reduce_sum.hpp

```diff
@@ -14,434 +14,240 @@
+  size_t count_var_impl(size_t count, const Container& x,
+                        const Pargs&... args) const {
+    return count_var_impl(count + x.size(), args...);
+  }
+
+  template <typename Container,
+            require_t<is_detected<Container, member_size_t>>...,
+            require_t<std::is_arithmetic<value_type_t<Container>>>...,
+            typename... Pargs>
+  size_t count_var_impl(size_t count, const Container& x,
+                        const Pargs&... args) const {
+    return count_var_impl(count, args...);
+  }
+
+  template <typename... Pargs>
+  size_t count_var_impl(size_t count, const var& x,
+                        const Pargs&... args) const {
+    return count_var_impl(count + 1, args...);
+  }
+
+  template <typename... Pargs, typename Arith,
+            require_arithmetic_t<scalar_type_t<Arith>>...>
+  size_t count_var_impl(size_t count, Arith& x, const Pargs&... args) const {
+    return count_var_impl(count, args...);
+  }
+
+  size_t count_var_impl(size_t count) const { return count; }
+
+  /**
+   * Count the number of scalars of type T in the input argument list
+   *
+   * @tparam Pargs Types of input arguments
+   * @return Number of scalars of type T in input
+   */
+  template <typename... Pargs>
+  size_t count_var(const Pargs&... args) const {
+    return count_var_impl(0, args...);
+  }
+
+  template <typename... Pargs>
+  void save_varis(vari** dest, const var& x, const Pargs&... args) const {
+    *dest = x.vi_;
+    save_varis(dest + 1, args...);
+  }
+
+  template <typename... Pargs, typename Vec,
+            require_vector_like_vt<is_var, Vec>...>
+  void save_varis(vari** dest, const Vec& x, const Pargs&... args) const {
+    for (size_t i = 0; i < x.size(); ++i) {
+      dest[i] = x[i].vi_;
+    }
+    save_varis(dest + x.size(), args...);
+  }
+
+  template <typename... Pargs, typename Mat,
+            require_t<is_detected<Mat, operator_paren_access_t>>...,
+            require_t<is_var<value_type_t<Mat>>>...>
+  void save_varis(vari** dest, const Mat& x, const Pargs&... args) const {
+    for (size_t i = 0; i < x.size(); ++i) {
+      dest[i] = x(i).vi_;
+    }
+    save_varis(dest + x.size(), args...);
+  }
+
+  template <typename R, typename... Pargs>
+  void save_varis(vari** dest, const R& x, const Pargs&... args) const {
+    save_varis(dest, args...);
+  }
+
+  void save_varis(vari**) const {}
+
+  var operator()(const std::vector<M>& vmapped, std::size_t grainsize,
+                 const Args&... args) const {
     const std::size_t num_jobs = vmapped.size();
 
     if (num_jobs == 0)
-      return init;
+      return var(0.0);
 
-    vmapped_op_t op_vmapped(vmapped);
+    const std::size_t num_terms = count_var(args...);
 
-    recursive_reducer worker(op_vmapped, num_elements(vmapped[0]), init, arg1,
-                             arg2, arg3, arg4);
+    recursive_reducer worker(num_terms, vmapped, args...);
 
 #ifdef STAN_DETERMINISTIC
```

Comment:
We should wrap up the parallel stuff so this macro is more behind the scenes

Reply:
I don't mind macros if these are only rare.

Reply:
personally I hate them being shoved into the middle of a function, we do that in the OpenCL stuff and it's gross. It doesn't help the readability at all and imo in this instance is tmi

Reply:
I'll leave this as optional but personally I'd like to have this pulled out into an internal function so we don't have a macro in the middle of a function that people need to split in their brains